### DISCUSSION TITLE
RBoot issue

### DISCUSSION THREAD

**Discussion by pljakobs** [SmingHub/Sming#2693] [General]
I have posted this over at the RBOOT repo, but maybe someone here has an idea? 
https://github.com/raburton/rboot/issues/64

pj

**Comment by mikee47** [comment]
For clarity, I've copied your original post here:

=========

I'm not sure this is an 'issue' the way it's meant, but I'm getting this a lot recently (using SMING):

```

 ets Jan  8 2013,rst cause:1, boot mode:(3,6)

load 0x40100000, len 1220, room 16
tail 4
chksum 0xda
load 0x3ffe8000, len 640, room 4
tail 12
chksum 0xa8
csum 0xa8

rBoot v1.4.2 - richardaburton@gmail.com
Flash Size:   32 Mbit
Flash Mode:   DIO
Flash Speed:  40 MHz

Resetting boot config.
Invalid rom selected, defaulting to 0.
Rom 0 at 0 is bad.
Rom -1 at 0 is bad.
Rom -1 at 0 is bad.
Rom -1 at 0 is bad.
```

What I have tried:

    * re-flashing the app - no change

    * make flashinit to clean the flash and then re-flashing the app - no change

    * try a brand new ESP8266 - no change in one case and success in another.


I have had odd issues of having to re-boot the esp numerous times but since I've mostly been using a serial speed of 460800baud, all I saw is unprintable characters filling the screen. I ssume this is an intermittend issue, sometimes clearing itself with a few re-boots, or rather, I should say, in some cases, it happens in four out of five boots, in other cases, it happens in very boot in a resonable number of tries.

I have mostly suspected worn out flash cells, but this latest occurence is, as said, on a fresh out of the package ESP. I have many more and will try others, but if yo have any other xplanation, I would be glad to hear that.

Oh, just for completenes: here's my hwconfig:

```
make hwconfig

esp_rgbww_firmware: Invoking 'hwconfig' for Esp8266 (debug) architecture

{
    "name": "Two ROM slots with single SPIFFS",
    "comment": "",
    "arch": "Esp8266",
    "options": [
        "4m",
        "spiffs"
    ],
    "bootloader_size": "8K",
    "partition_table_offset": "0x003fa000",
    "devices": {
        "spiFlash": {
            "type": "flash",
            "size": "4M",
            "mode": "dio",
            "speed": 40
        }
    },
    "partitions": {
        "rom0": {
            "device": "spiFlash",
            "address": "0x00002000",
            "size": "992K",
            "type": "app",
            "subtype": "ota_0",
            "readonly": false,
            "encrypted": false,
            "filename": "$(RBOOT_ROM_0_BIN)",
            "unused_before": 0,
            "unused_after": 0
        },
        "rom1": {
            "device": "spiFlash",
            "address": "0x00102000",
            "size": "992K",
            "type": "app",
            "subtype": "ota_1",
            "readonly": false,
            "encrypted": false,
            "filename": "$(RBOOT_ROM_0_BIN)",
            "unused_before": 0,
            "unused_after": 0
        },
        "spiffs0": {
            "device": "spiFlash",
            "address": "0x00200000",
            "size": "512K",
            "type": "data",
            "subtype": "spiffs",
            "readonly": false,
            "encrypted": false,
            "filename": "$(SPIFF_BIN_OUT)",
            "build": {
                "target": "spiffsgen",
                "files": "$(SPIFF_FILES)"
            },
            "unused_before": 0,
            "unused_after": 0
        },
        "rf_cal": {
            "device": "spiFlash",
            "address": "0x003fb000",
            "size": "4K",
            "type": "data",
            "subtype": "rfcal",
            "readonly": false,
            "encrypted": false,
            "filename": "",
            "unused_before": 0,
            "unused_after": 0
        },
        "phy_init": {
            "device": "spiFlash",
            "address": "0x003fc000",
            "size": "4K",
            "type": "data",
            "subtype": "phy",
            "readonly": false,
            "encrypted": false,
            "filename": "$(FLASH_INIT_DATA)",
            "unused_before": 0,
            "unused_after": 0
        },
        "sys_param": {
            "device": "spiFlash",
            "address": "0x003fd000",
            "size": "12K",
            "type": "data",
            "subtype": "sysparam",
            "readonly": false,
            "encrypted": false,
            "filename": "",
            "unused_before": 0,
            "unused_after": 0
        }
    }
}
```

**Comment by mikee47** [comment]
Please note that Sming uses a fork of rBoot (#2258) located here https://github.com/mikee47/rboot/tree/sming. This was done to accommodate partition tables.

I've flashed the `Basic_Ota` sample to an ESP-12F for comparison: this uses a similar hardware config to yours.

```
 ets Jan  8 2013,rst cause:2, boot mode:(3,7)

load 0x40100000, len 1220, room 16 
tail 4
chksum 0xda
load 0x3ffe8000, len 640, room 4 
tail 12
chksum 0xa8
csum 0xa8

rBoot v1.4.2 - richardaburton@gmail.com
Flash Size:   32 Mbit
Flash Mode:   DIO
Flash Speed:  40 MHz

Found 'rom0' @ 0x00002000, size 0x000f8000, subtype 0x10 ///< Missing
Found 'rom1' @ 0x00102000, size 0x000f8000, subtype 0x11 ///< Missing
Booting rom 0 at 2000, load addr 4e200.
```

You can follow the logic here https://github.com/mikee47/rboot/blob/cc00071f593ef7d55c0c0a7b97828173f9cda89b/rboot.c#L392.

Comparing this log to yours I can make two observations:

First, the `Resetting boot config` message indicates that the rBoot config is corrupt (or has been wiped). You may see this once after running `flashinit` but in normal operation this shouldn't happen.

Second, the subsequent call to `scan_partitions` yields no results, which is why the device can't boot. Possible explanations:

1. The partition table is corrupted. Run `make readmap` to read the partition table off the device and display it, for comparison with `make map`. You can use `make verifyflash` to read back and verify all partitions (and the partition table).
2. rBoot is looking for the partition table in the wrong place. Try rebuilding and re-flashing just rBoot, `make rebuild; make flashboot`.

However, either of these I would expect permanent boot failure, not the erratic issue you describe. My first though is to check you have a stable power supply with adequate decoupling next to the esp8266. If the problem is reproduceable then perhaps adding debug statements to rboot might help?

**Comment by mikee47** [comment]
Another thought, does your application use OTA updates? If so, that could be implicated...

**Comment by pljakobs** [comment]
sorry, I never came back to this, I resolved it somehow, can't really remember. 
I will have some more OTA / rBoot / partition table questions, but for those, I'll open a new discussion.
